### ISSUE TITLE
Email tracking link becomes "404 not found" if order is resaved after it was sent

### ISSUE THREAD

**Issue by mattias** [magento/magento2#30239]
### Preconditions (*)

1. Magento 2.3.5-p2
2. Any shipping service that provides a tracking number and link. 

### Steps to reproduce (*)

1. Navigate to storefront and place an order.
2. Go to admin and create a shipment, as well as checking "Create Shipping Label" and "Email Copy of Shipment".
3. If you look at the email copy sent out now, the tracking link will work.
4. Now create an invoice for the order.

### Expected result (*)

1. The tracking link should still work in the email that was sent out eariler.

### Actual result (*)

1. The tracking link becomes "404 not found" after the order has become invoiced.

### Information
We've investigated this a little and it comes down to the code in vendor/magento/module-sales/Model/ResourceModel/Order.php:156. It uses a random number each time to create the hash right before the order is saved. Apparently you have to invoice first and ship after for this not to happen. Our client however ships first then creates the invoice.

Does this piece of code have to be regenerated on each save of the order? Would it be possible to check if it already exists before trying to add a new random code?

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [X] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by ihorvansach** [comment]
@magento give me 2.4-develop instance

**Comment by ihorvansach** [comment]
I can confirm the issue. It is related to the order field "protect_code". 
It is in use for creating shipping/tracking/popup page hash key and it is changing each time during order save (reference \Magento\Sales\Model\ResourceModel\Order::_beforeSave), e.g. crating an invoice.

Think need to add some check to do not change it so often, e.g. not earlier than once a month.

**Comment by ihorvansach** [comment]
UPD: the issue is present in the Magento 2.4.1 but it is already fixed in the 2.4-develop branch.
Here is the commit: https://github.com/magento/magento2/commit/bd087ecd950024190fef4e5cb6fe5aa6bb27d2eb

To solve the issue on your instanse you can apply a patch

**Patch**
MDVA-30265: Fixes the issue where shipment tracking link stops working a filter Invoice creation.

**Compatible versions**
Magento OpenSource/Commerce/Commerce Cloud 2.3.3 - 2.4.1

**warning** We strongly recommend testing all patches in a staging or development environment before deploying to production.

**Applying a patch - Magento OpenSource/Commerce**

```
$ composer require magento/quality-patches
$ ./vendor/bin/magento-patches apply MDVA-30265
```
See [MQP Magento Commerce documentation](https://devdocs.magento.com/guides/v2.4/comp-mgr/patching/mqp.html)

Applying a patch - Magento Commerce Cloud
See [MQP Magento Commerce Cloud documentation](https://devdocs.magento.com/cloud/project/project-patch.html)

Patch Details
https://support.magento.com/hc/en-us/articles/360049311912
